### PR TITLE
✨ clusterctl alpha rollout pause/resume for MachineDeployments

### PR DESCRIPTION
**Pull request by Arvinderpal**
**What this PR does / why we need it**:
Adds command and client for the `clusterctl alpha rollout pause/resume`. 

Tracking Issue: 
https://github.com/kubernetes-sigs/cluster-api/issues/3439

### CODE REVIEW

**Comment by Arvinderpal** [comment]
@wfernandes @vincepri PTAL. 
I also have the `undo` command ready to go but will wait until this is merged. 
Thank you

**Comment by wfernandes** [comment]
@Arvinderpal Thanks for this PR! I'll take a look at this first thing next week 🙂

File: cmd/clusterctl/client/rollout.go
Comment (wfernandes) [inline cmd/clusterctl/client/rollout.go:94]: This code seems duplicated in every method. I suggest refactoring it to consolidate this.

File: cmd/clusterctl/client/rollout_test.go
Comment (wfernandes) [inline cmd/clusterctl/client/rollout_test.go:35]: Let's refactor this test so that we can assert on each of the rollout behaviors - pause, resume, restart.

File: cmd/clusterctl/cmd/rollout/pause.go
Comment (wfernandes) [inline cmd/clusterctl/cmd/rollout/pause.go:39]: Pulled this from the `kubectl rollout pause -h`
```suggestion
		Mark the provided cluster-api resource as paused.

	        Paused resources will not be reconciled by a controller. Use "clusterctl alpha rollout resume" to resume a paused resource. Currently only MachineDeployments support being paused.`)
```

File: cmd/clusterctl/cmd/rollout/resume.go
Comment (wfernandes) [inline cmd/clusterctl/cmd/rollout/resume.go:40]: Again following the pattern of `kubectl rollout resume -h`.
```suggestion
		Resume a paused cluster-api resource

                 Paused resources will not be reconciled by a controller. By resuming a resource, we allow it to be reconciled again.
Currently only MachineDeployments support being resumed.`)
```

Also I noticed the indentation was off when I printed out the help for this comamand.
```
 $ clusterctl alpha rollout resume -h
Resume of cluser-api paused resources.

     Paused resources will not be reconciled by a controller. By resuming a

 resource, we allow it to be reconciled again.
```

File: cmd/clusterctl/cmd/rollout/pause.go
Comment (wfernandes) [inline cmd/clusterctl/cmd/rollout/pause.go:69]: From the linter, since we are not using `cmd *cobra.Command` in this method, we can remove it from the argument list.

**Review by wfernandes — COMMENTED** [review]
Thanks for the PR @Arvinderpal. Just some minor comments 🙂

File: cmd/clusterctl/client/rollout.go
Comment (Arvinderpal) [reply cmd/clusterctl/client/rollout.go:94]: Refactored it.

File: cmd/clusterctl/client/rollout_test.go
Comment (Arvinderpal) [reply cmd/clusterctl/client/rollout_test.go:35]: Done. PTAL

File: cmd/clusterctl/cmd/rollout/resume.go
Comment (Arvinderpal) [reply cmd/clusterctl/cmd/rollout/resume.go:40]: Done.

**Review by Arvinderpal — COMMENTED** [review]
@wfernandes Thank you for the review. I think I have addressed all your comments.

File: cmd/clusterctl/client/rollout.go
Comment (wfernandes) [inline cmd/clusterctl/client/rollout.go:48]: Umm...is this correct? Or should this be `return err`? It also looks like the tests all passed having this logic in so if this is indeed meant to be `return err` then we should have a unit test that asserts this. Thanks!

File: cmd/clusterctl/client/rollout.go
Comment (wfernandes) [inline cmd/clusterctl/client/rollout.go:65]: Same here.

File: cmd/clusterctl/client/rollout.go
Comment (wfernandes) [inline cmd/clusterctl/client/rollout.go:82]: Same here.

**Review by wfernandes — COMMENTED** [review]
Just a clarification. Thanks.

File: cmd/clusterctl/client/rollout.go
Comment (Arvinderpal) [reply cmd/clusterctl/client/rollout.go]: Thanks for pointing that out! The fakeClient was not calling the individual funcs for the subcommands. 
I also had to separate these tests -- for example, the RolloutResume will throw an error if the md is not paused, while the RolloutRestart will do the opposite if it is paused, so the same test cases cannot be passed to all of them. 
There are additional tests in the lower level funcs that check for functional correctness. Is it okay if we keep these tests relatively simple?

Comment (wfernandes) [reply cmd/clusterctl/client/rollout.go]: yeah...that's fine. I just wanted to ensure that if I changed this to `return nil` and run the tests in the `client` pkg, then I would see a failing test for the right reason. 🙂

**Comment by wfernandes** [comment]
/lgtm
Over to @fabriziopandini or @CecileRobertMichon for approval.

**Comment by Arvinderpal** [comment]
Thanks @wfernandes 
@fabriziopandini @CecileRobertMichon PTAL.
I have the `undo` command PR ready to go after this one as well.

**Comment by wfernandes** [comment]
@Arvinderpal If you'd like to submit the PR for the `undo` rollout command we can temporarily put a WIP or a `/hold` until this gets merged and then we can rebase onto it. 
Just a suggestion so you can get eyes on the other PR immediately instead of having to wait for this to get merged. 🙂

**Comment by Arvinderpal** [comment]
> @Arvinderpal If you'd like to submit the PR for the `undo` rollout command we can temporarily put a WIP or a `/hold` until this gets merged and then we can rebase onto it.
> Just a suggestion so you can get eyes on the other PR immediately instead of having to wait for this to get merged.

@wfernandes That's a good idea. I'll open an `undo` PR with a wip on it.

File: cmd/clusterctl/client/alpha/rollout_resumer.go
Comment (vincepri) [inline cmd/clusterctl/client/alpha/rollout_resumer.go:38]:
```suggestion
			return errors.Errorf("MachineDeployment is not currently paused: %v/%v\n", tuple.Resource, tuple.Name)
```

File: cmd/clusterctl/client/alpha/rollout_resumer.go
Comment (vincepri) [inline cmd/clusterctl/client/alpha/rollout_resumer.go:44]:
```suggestion
		return errors.Errorf("Invalid resource type %q, valid values are %v", tuple.Resource, validResourceTypes)
```

File: cmd/clusterctl/client/alpha/rollout_pauser.go
Comment (vincepri) [inline cmd/clusterctl/client/alpha/rollout_pauser.go:38]:
```suggestion
			return errors.Errorf("MachineDeployment is already paused: %v/%v\n", tuple.Resource, tuple.Name)
```

**Comment by Arvinderpal** [comment]
@vincepri Can we merge this?

File: cmd/clusterctl/cmd/rollout/pause.go
Comment (CecileRobertMichon) [inline cmd/clusterctl/cmd/rollout/pause.go:48]: are you planning to add docs for the pause/resume command in a separate PR?

Comment (CecileRobertMichon) [reply cmd/clusterctl/cmd/rollout/pause.go:48]: I actually don't see any docs for the rollout command in https://github.com/kubernetes-sigs/cluster-api/tree/master/docs/book/src/clusterctl/commands, are we tracking that in an issue?

Comment (Arvinderpal) [reply cmd/clusterctl/cmd/rollout/pause.go:48]: @CecileRobertMichon There is a tracking issue and docs is in the list: #3439
If it's okay with you, I want to wait for a few more rollout subcommands to merge before opening a PR for the docs.

**Review by vincepri — APPROVED** [review]
/approve
/hold 
/assign @CecileRobertMichon 
to remove the hold if everything looks 👍

**Review by CecileRobertMichon — COMMENTED** [review]
/hold cancel

**Pull request merged**
